### Задача 1



Машинным $ε$ называется такое число, что $1 + ε/2 = 1$, но $1 + ε \neq 1$. (Также часто
используется обозначение ULP – unit in the last place, или unit of least precision, единица
в младшем разряде.) Найти машинное $ε$, число разрядов в мантиссе, максимальную
и минимальную степени, при вычислениях с обычной и двойной точностью. Сравнить
друг с другом четыре числа: $1, 1 + ε/2, 1 + ε, 1 + ε + ε/2$ , объяснить результат.

#### Теория: документация и методичка

Floating point numbers are usually implemented using double in C; information about the precision and internal representation of floating point numbers for the machine on which your program is running is available in sys.float_info.

In [42]:
import sys
import math

In [22]:
#будет, с чем свериться!
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

In [5]:
0.1+0.2-0.3

5.551115123125783e-17

Результат произвольной арифметической операции также может содержать относительную погрешность округления $2 −(𝑛+1) = 𝜀/2$. Число $𝜀$ называют машинным эпсилон, или ULP, единицей в младшем разряде мантиссы. Другими словами, число $𝜀$ — это минимальная степень двойки (или, в общем случае, основания системы счисления), при сложении которой с единицей результат будет отличен от единицы.

$𝜀$ характеризует именно относительную погрешность; абсолютная величина ошибки равна произведению $1/2𝜀 \cdot 2^𝐸$, где 𝐸 — двоичный порядок вычисляемой величины. Указанная относительная ошибка округления ($𝜀/2$) относится к операциям умножения и деления, а также возникает при вводе числовых констант.

**Следовательно из (\*):  $𝜀 = 2^{-n}$**

**Как буду искать $E_{max}$ :**

$2^{E_{max}} \cdot 2 = \infty$

$E_{max} \rightarrow  E_{s max} = 11....10 = 2^\omega - 1$

In [62]:
print(float('inf') > 2*1.7976931348623157e+308) 
print(2*1.7976931348623157e+308 < 4*1.7976931348623157e+308) 

#эксперименты для проверки факта, что все бесконечности равны

False
False


### шаг 1: найду 𝜀 и n

$x = 2^{-n}$

цикл остановится при $x = \epsilon/2$

In [52]:
x = 1
n = 0

while (x + 1 != 1):
    x = x/2
    n = n+1
    
n = n - 1 
epsilon = x*2
print("epsilon = ", epsilon)
print("mantissa digits n = ", n)

epsilon =  2.220446049250313e-16
mantissa digits n =  52


### шаг 2: найду Emax

на старте $x_{current} = 1$

в цикле 

$x_{current} = (-1)^0 \cdot (1 + 0) \cdot 2^E = 2^E$

$x_{previous} = 2^{E-1}$

In [29]:
#не рабочая попытка 1
x_current = 2
x_previous = 1
E = 1

while(x_previous < x_current):
    E = E+1
    x_previous = 2**(E-1)
    x_current = 2**E
    
print("E max = ", E)

KeyboardInterrupt: 

In [34]:
#не рабочая попытка 2
x_current = 2
E = 1

while(2**(E-1) < 2**E):
    E = E+1
    
E_max = E - 1
print("E max = ", E)

KeyboardInterrupt: 

In [46]:
#внезапно рабочая попытка

x = 1.
E = 0

while(not math.isinf(x)):
    x = x*2
    E = E+1
    
E_max = E - 1
print("E max = ", E_max)

E max =  1023


### шаг 3: найду Emin через w

$E_{max} = -1 + 2^{\omega - 1}$

$E_{min} = -2^{\omega - 1} + 2$

In [49]:
w = math.log(E_max+1, 2) + 1
E_min = -1 * 2**(w - 1) + 2

print("w =", int(w))
print("E min =", int(E_min))

w =  11
E min =  -1022


### Результаты

число битов на дробную часть мантиссы $n = 52$  *совпало с методичкой, не совпало с sys.float_info*

машинная $\epsilon = 2.220446049250313e-16$  *совпало с методичкой, совпало с sys.float_info*

число битов на показатель степени 2 $\omega = 11$  *совпало со здравым смыслом: 64 - 52 мантисса - 1 на знак = 11*

максимальная степень $E_{max} = 1023$ *не совпало с sys.float_info*

минимальная степень $E_{min} = -1022$  *не совпало с sys.float_info*

### Сравнение чисел

In [53]:
print(1 < 1 + epsilon/2)
print(1 + epsilon/2 < 1 + epsilon)
print(1 + epsilon < 1 + epsilon + epsilon/2)

False
True
True


$1 = 1 + ε/2 < 1 + ε < 1 + ε + ε/2$

In [61]:
import numpy as np